### ISSUE TITLE
Rule suggestion: restrict in operator

### ISSUE THREAD

**Issue by denis-sokolov** [eslint/eslint#2723]
I would like to propose a warning for `bar in foo` where `bar` is a variable.
I might be mistaken, but it's likely the author means `Object.prototype.hasOwnProperty.call(foo, bar)`.
In cases where the user actually wants to check for an inherited property existence, such as feature detects, usually the value is a constant: `if ('quux' in foo) f(foo.quux)`.
eslint codebase has 0 occurences of an `in` operator with a variable as the first operand.

The following examples would cause a warning:
```javascript
if (key in myDictionary) {
}

var isWordMentioned = function(word){
  return word in mentionedWords;
};
```

The following examples would not cause anything:
```javascript
for (var key in obj) {
}

var userWantsCookies = 'cookies' in userWishes;
```

Ultimately I want to catch problems caused by using `{}` as maps where `Object.create(null)` or `new Map()` would be more fitting.

Feedback is appreciated.
Tangentially related to #2693 as it's all caused by JS objects used as maps.

**Comment by ilyavolodin** [comment]
I'm not 100% sure I follow what you are asking for. Could you provide a few valid and invalid use cases?

**Comment by denis-sokolov** [comment]
Sorry about that. I have updated the main comment with examples.

**Comment by nzakas** [comment]
`in` searches the entire prototype chain, not just the immediate prototype, so these are not identical.

From the description, this seems more like a personal preference than a generalizable rule, so I think it would be better off as a custom rule.

**Comment by denis-sokolov** [comment]
@nzakas, the entire point of my rule is these are not identical.
The `in` in cases presented is a programmer's logical error and that's what I am trying to prevent.

**Comment by nzakas** [comment]
Yes, I just think it's a huge leap to say using `in` is probably a mistake. 

Remember, the whole point of allowing custom rules is so you don't have to convince us of your point of view. You can still write a rule to do this, I just dying think it fits as a core rule

**Comment by denis-sokolov** [comment]
Sure, if you don't think it fits a core rule, I got my answer.

Your previous answer seemed misled by me not explaining myself well enough, thus I have asked for clarification. Thanks!
